File: domains/finance/projects/stock-market/development/software/ios/StockPrediction/Sync/SyncManager.swift
import Foundation

actor SyncManager {
    static let shared = SyncManager()

    // Tickers synced in "Quick Sync" (always included)
    private let hotTickers = [
        "NVDA","AAPL","MSFT","META","GOOGL","AMZN","TSLA","AMD","NFLX",
        "HOOD","PLTR","ARM","SMCI","COIN","MSTR","UBER","LYFT","SOFI",
        "RBLX","SNAP","RIVN","SOUN","AI","IONQ","QUBT","RDDT","ACHR","JOBY",
    ]

    // Top 50 S&P 500 by market cap (hardcoded for reliability)
    private let sp500Top50 = [
        "AAPL","MSFT","NVDA","AMZN","GOOGL","META","TSLA","BRK-B","AVGO","JPM",
        "LLY","V","XOM","UNH","MA","JNJ","HD","PG","COST","MRK",
        "ABBV","BAC","WMT","CVX","NFLX","KO","AMD","PEP","TMO","ORCL",
        "CSCO","CRM","ABT","ACN","MCD","ADBE","TXN","PM","DHR","VZ",
        "CAT","WFC","INTU","SPGI","NOW","NEE","AXP","UPS","MS","DE",
    ]

    enum SyncError: LocalizedError {
        case noData(String)
        case networkError(String)

        var errorDescription: String? {
            switch self {
            case .noData(let t):       return "No data returned for \(t)"
            case .networkError(let m): return "Network error: \(m)"
            }
        }
    }

    // Progress callback: (tickersDone, totalTickers, currentTicker)
    func sync(fullSP500: Bool = false, onProgress: @escaping (Int, Int, String) -> Void) async throws -> String {
        var tickers: [String]
        if fullSP500 {
            tickers = Array(Set(hotTickers + sp500Top50)).sorted()
        } else {
            tickers = Array(Set(hotTickers + sp500Top50)).sorted()
        }
        let total = tickers.count
        var done = 0

        for ticker in tickers {
            onProgress(done, total, ticker)
            do {
                let bars = try await fetchBars(ticker: ticker)
                if !bars.isEmpty {
                    try DatabaseManager.shared.upsertPrices(bars)
                }
                // Also fetch/update stock info for first-time tickers
                if let stock = try? await YahooFinanceFetcher.fetchQuote(ticker: ticker) {
                    try? DatabaseManager.shared.upsertStocks([stock])
                }
                // Run prediction and cache
                if let pred = try? PredictionEngine.shared.predict(ticker: ticker, horizon: "1w") {
                    try? DatabaseManager.shared.upsertPrediction(pred)
                }
            } catch {
                // Skip failed tickers — don't abort the whole sync
            }
            done += 1
            // Rate limit: 0.3s between calls to avoid 429
            try? await Task.sleep(nanoseconds: 300_000_000)
        }
        onProgress(total, total, "")
        return "Synced \(done) tickers"
    }

    private func fetchBars(ticker: String) async throws -> [PriceBar] {
        switch MarketDataSettings.activeSource {
        case .yahooFinance:
            return try await YahooFinanceFetcher.fetchPriceBars(ticker: ticker)
        case .stooq:
            return try await StooqFetcher.fetchPriceBars(ticker: ticker)
        case .alphaVantage, .twelveData, .polygonIO:
            // Fallback to Yahoo if no API key configured
            let key = MarketDataSettings.apiKey(for: MarketDataSettings.activeSource)
            if key.isEmpty { return try await YahooFinanceFetcher.fetchPriceBars(ticker: ticker) }
            // TODO: implement keyed sources — fall back to Yahoo for now
            return try await YahooFinanceFetcher.fetchPriceBars(ticker: ticker)
        }
    }
}

File: domains/finance/projects/stock-market/development/software/android/core/src/main/java/com/prediction/stockmarket/data/sync/SyncManager.kt
package com.prediction.stockmarket.data.sync

import com.prediction.stockmarket.data.database.*
import com.prediction.stockmarket.prediction.PredictionEngine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SyncManager @Inject constructor(
    private val stockDao: StockDao,
    private val priceDao: PriceDao,
    private val predictionDao: PredictionDao,
    private val yahooFetcher: YahooFinanceFetcher,
    private val stooqFetcher: StooqFetcher,
    private val sourceManager: MarketDataSourceManager,
    private val predictionEngine: PredictionEngine
) {
    sealed class SyncResult {
        data class Success(val tagName: String) : SyncResult()
        data class Error(val message: String) : SyncResult()
    }

    private val hotTickers = listOf(
        "NVDA", "AAPL", "MSFT", "META", "GOOGL", "AMZN", "TSLA", "AMD", "NFLX",
        "HOOD", "PLTR", "ARM", "SMCI", "COIN", "MSTR", "UBER", "LYFT", "SOFI",
        "RBLX", "SNAP", "RIVN", "SOUN", "AI", "IONQ", "QUBT", "RDDT", "ACHR", "JOBY"
    )

    private val sp500Top50 = listOf(
        "AAPL", "MSFT", "NVDA", "AMZN", "GOOGL", "META", "TSLA", "AVGO", "JPM", "LLY",
        "V", "XOM", "UNH", "MA", "JNJ", "HD", "PG", "COST", "MRK", "ABBV",
        "BAC", "WMT", "CVX", "NFLX", "KO", "AMD", "PEP", "TMO", "ORCL", "CSCO",
        "CRM", "ABT", "ACN", "MCD", "ADBE", "TXN", "PM", "DHR", "VZ", "CAT",
        "WFC", "INTU", "SPGI", "NOW", "NEE", "AXP", "UPS", "MS", "DE", "GE"
    )

    suspend fun sync(
        onProgress: (done: Int, total: Int, currentTicker: String) -> Unit = { _, _, _ -> }
    ): SyncResult = withContext(Dispatchers.IO) {
        try {
            val tickers = (hotTickers + sp500Top50).distinct()
            val total = tickers.size
            val source = sourceManager.activeSource
            var successCount = 0

            tickers.forEachIndexed { index, ticker ->
                onProgress(index, total, ticker)
                try {
                    // Fetch price bars from the active source
                    val priceBars = when (source) {
                        MarketDataSourceType.STOOQ -> stooqFetcher.fetchPriceBars(ticker)
                        else -> yahooFetcher.fetchPriceBars(ticker)
                    }

                    if (priceBars.isNotEmpty()) {
                        // Fetch and upsert stock metadata
                        val stockEntity = yahooFetcher.fetchQuote(ticker)
                            ?: com.prediction.stockmarket.data.database.StockEntity(
                                ticker = ticker,
                                name = ticker
                            )
                        stockDao.upsertAll(listOf(stockEntity))

                        // Upsert price bars
                        priceDao.upsertAll(priceBars)

                        // Run prediction engine for each horizon
                        val sortedBars = priceBars.sortedByDescending { it.date }
                        listOf("1W", "1M", "3M").forEach { horizon ->
                            val prediction = predictionEngine.predict(ticker, horizon, sortedBars)
                            if (prediction != null) {
                                predictionDao.upsertAll(listOf(prediction))
                            }
                        }

                        successCount++
                    }
                } catch (_: Exception) {
                    // Skip failed tickers and continue
                }

                // Rate limit: 300ms between calls
                delay(300L)
            }

            onProgress(total, total, "")
            SyncResult.Success("Synced $successCount tickers from ${source.displayName}")
        } catch (e: Exception) {
            SyncResult.Error(e.message ?: "Unknown error")
        }
    }
}
